combin/combin.go

```diff
@@ -0,0 +1,187 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package combin implements routines involving combinatorics (permutations,
+// combinations, etc.).
+package combin
+
+import "math"
+
+const (
+	badNegInput = "combin: negative input"
+	badSetSize  = "combin: n < k"
+	badInput    = "combin: wrong input slice length"
+)
+
+// Binomial returns the binomial coefficient of (n,k), also commonly referred to
+// as "n choose k".
+//
+// The binomial coefficient, C(n,k), is the number of unordered combinations of
+// k elements in a set that is n elements big, and is defined as
+//
+//  C(n,k) = n!/((n-k)!k!)
+//
+// n and k must be non-negative with n >= k, otherwise Binomial will panic.
+// No check is made for overflow.
+func Binomial(n, k int) int {
+	if n < 0 || k < 0 {
+		panic(badNegInput)
+	}
+	if n < k {
+		panic(badSetSize)
+	}
+	// (n,k) = (n, n-k)
+	if k > n/2 {
+		k = n - k
+	}
+	kfact := 1
+	for i := 2; i <= k; i++ {
+		kfact *= i
+	}
+	num := 1
+	for i := n; i > n-k; i-- {
+		num *= i
+	}
+	return num / kfact
+}
+
+// GeneralizedBinomial returns the generalized binomial coefficient of (n, k),
+// defined as
+//  Γ(n+1) / (Γ(k+1) Γ(n-k+1))
+// where Γ is the Gamma function. GeneralizedBinomial is useful for continuous
+// relaxations of the binomial coefficient, or when the binomial coefficient value
+// may overflow int. In the latter case, one may use math/big for an exact
+// computation.
+//
+// n and k must be non-negative with n >= k, otherwise GeneralizedBinomial will panic.
+func GeneralizedBinomial(n, k float64) float64 {
+	return math.Exp(LogGeneralizedBinomial(n, k))
+}
+
+// LogGeneralizedBinomial returns the log of the generalized binomial coefficient.
+// See GeneralizedBinomial for more information.
+func LogGeneralizedBinomial(n, k float64) float64 {
+	if n < 0 || k < 0 {
+		panic(badNegInput)
+	}
+	if n < k {
+		panic(badSetSize)
+	}
+	a, _ := math.Lgamma(n + 1)
+	b, _ := math.Lgamma(k + 1)
+	c, _ := math.Lgamma(n - k + 1)
+	return a - b - c
+}
+
+// CombinationGenerator generates combinations iteratively. Combinations may be
+// called to generate all combinations collectively.
+type CombinationGenerator struct {
+	n         int
+	k         int
+	previous  []int
+	remaining int
+}
+
+// NewCombinationGenerator returns a CombinationGenerator for generating the
+// combinations of k elements from a set of size n.
+//
+// n and k must be non-negative with n >= k, otherwise NewCombinationGenerator
+// will panic.
+func NewCombinationGenerator(n, k int) *CombinationGenerator {
```

Comment:
(apologies for joining late)
any reason why you return a pointer?
it's still reasonnably small (3ints+3words)...